cdap-app-templates/cdap-etl/hydrator-spark-core-base/src/main/java/io/cdap/cdap/etl/spark/batch/OpaqueDatasetCollection.java

```diff
@@ -108,12 +110,44 @@ public SparkCollection<T> join(JoinExpressionRequest joinRequest) {
   @Override
   public DataframeCollection toDataframeCollection(Schema schema) {
     StructType sparkSchema = DataFrames.toDataType(schema);
-    ExpressionEncoder<Row> encoder = RowEncoder.apply(sparkSchema);
+    ExpressionEncoder<Row> encoder = getRowEncoder(sparkSchema);
     Dataset<StructuredRecord> ds = (Dataset<StructuredRecord>) getDataset();
     MapFunction<StructuredRecord, Row> converter = r -> DataFrames.toRow(r, sparkSchema);
     return new DataframeCollection(schema, ds.map(converter, encoder),
         sec, jsc, sqlContext, datasetContext, sinkFactory, functionCacheFactory);
   }
 
+  /**
+   * This is required to handle breaking changes between spark 3.3.2 (Dataproc 2.1) to 3.5.1 (Dataproc 2.2).
+   * And we need to support both.
+   * Here we are trying to check if the new method introduced in 3.5.1 exists or not, and based on that we
+   * invoke the new method or the old one.
+   */
+  public static ExpressionEncoder<Row> getRowEncoder(StructType sparkSchema) {
```

Comment:
Could you please cache the Method objects retrieved? Instrospection is a pretty heavy operation, we can do it once and then simply do "invoke".